app/frontend/components/common/asset.tsx

```diff
@@ -68,57 +85,122 @@ export const FormattedAssetItem = ({
+    const {labelType, label, labelShort} = (() => {
+      if (type === AssetFamily.ADA) {
+        const label = <span className={styles.assetName}>ADA</span>
+        return {
+          labelType: FormattedHumanReadableLabelType.ADA,
+          label,
+          labelShort: label,
+        }
+      }
 
-  const FormattedAssetName = () => {
-    if (type === AssetFamily.ADA) return <span className={styles.assetName}>ADA</span>
-    return (
-      <span>
-        {
-          <span className={metadata || assetName ? styles.assetName : styles.empty}>
-            {formatAssetName(assetName, metadata)}
-          </span>
+      if (metadata?.name) {
+        return {
+          labelType: FormattedHumanReadableLabelType.OFFCHAIN_ASSET_NAME,
+          label: (
+            <span className={styles.assetName}>
+              {`${metadata.name}${metadata?.ticker ? ` (${metadata.ticker})` : ''}`}
+            </span>
+          ),
+          labelShort: <span className={styles.assetName}>{metadata.ticker}</span>,
+        }
+      }
+
+      const readableAssetName = assetName && assetNameHex2Readable(assetName)
+      if (assetName && isHumanReadable(readableAssetName)) {
+        return {
+          labelType: FormattedHumanReadableLabelType.ONCHAIN_ASSET_NAME,
+          label: <span className={styles.assetName}>{readableAssetName}</span>,
+          labelShort: (
+            <span className={`${styles.assetName} flex-nowrap shrinkable`}>
+              <div className="ellipsis">{readableAssetName}</div>
+            </span>
+          ),
         }
-        <span className={styles.assetName}>{metadata?.ticker ? ` (${metadata?.ticker})` : ''}</span>
-      </span>
+      }
+
+      const label = (
+        <span className={`${styles.assetName} flex-nowrap shrinkable`}>
+          <StringEllipsis value={fingerprint!} length={6} />
+        </span>
+      )
+      return {
+        labelType: FormattedHumanReadableLabelType.FINGERPRINT,
+        label,
+        labelShort: label,
+      }
+    })()
+
+    const LabelWrapper = ({children}: {children: h.JSX.Element | h.JSX.Element[]}) => (
+      <div
+        className={`${styles.iconName} ${
+          labelType === FormattedHumanReadableLabelType.FINGERPRINT ? 'flex-nowrap shrinkable' : ''
+        }`}
+      >
+        {children}
+      </div>
     )
-  }
 
-  const FormattedAmount = () => {
-    if (type === AssetFamily.TOKEN) {
-      return printTokenAmount(quantity, metadata?.decimals || 0)
-    } else {
-      return printAda(Math.abs(quantity) as Lovelace)
+    return {
+      labelType,
+      label: <LabelWrapper>{label}</LabelWrapper>,
+      labelWithIcon: (
+        <LabelWrapper>
+          <div className={`${styles.icon} ${styles.iconOffset}`}>{icon}</div>
+          {label}
+        </LabelWrapper>
+      ),
+      shortLabelWithIcon: (
+        <LabelWrapper>
+          <div className={`${styles.icon} ${styles.iconOffset}`}>{icon}</div>
+          {labelShort ?? label}
```

Comment:
looking at the code above, isn't `labelShort` always defined, making the null-coalescing here redundant?

```suggestion
          {labelShort}
```

Reply:
Refactored